### PR TITLE
🔥 Stop syncing the value attribute on inputs (behind a feature flag)

### PR DESCRIPTION
**Pull request by nhunzaker**
This PR revives https://github.com/facebook/react/pull/10150, my original PR to remove value attribute syncing.

**TODOs**:

- [x] value set value, defaultValue set defaultValue
- [x] checked sets checked, defaultChecked sets defaultChecked
- [x] Remove warning if defaultValue changes (if it exists)
- [x] Remove warning if defaultChecked changes (if it exists)
- [ ] Can we get rid of wrapper state alltogether?
- [x] Could we only apply value to text inputs?
- [x] Could we apply checked to check/radios?
- [ ] Could submit/reset just use standard attribute assignment?
- [ ] SSR mismatch?

This fixture build demonstrates the behavior:

http://react-dom-no-sync-value.surge.sh/

---

See https://github.com/facebook/react/issues/11896 and https://github.com/facebook/react/issues/13525

🔥 🔥 🔥

### CODE REVIEW

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js [hunk 1]
```diff
@@ -1605,15 +1605,15 @@ describe('ReactDOMInput', () => {
       };
     }
 
-    it('always sets the attribute when values change on text inputs', function() {
+    it('retains the initial value attribute when values change on text inputs', function() {
```
Comment (nhunzaker) [inline packages/react-dom/src/__tests__/ReactDOMInput-test.js]: It's always been this way, but I don't like how React stores an initial value:
https://github.com/facebook/react/blob/master/packages/react-dom/src/client/ReactDOMFiberInput.js#L25

I think the value should just be blank when it becomes null, or revert to the defaultValue prop (which could be null).

**Comment by react-linter** [comment]
ReactDOM: size: -0.2%, gzip: -0.2%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 0452c9bba5707e4473fff17a24d36402a957dc66...10473d0619171d9be400b633b6b9f3e9605b0c3a</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | -0.0% | -0.4% | 660.24 KB | 660.03 KB | 154.94 KB | 154.29 KB | UMD_DEV
**react-dom.production.min.js** | **-0.2%** | **-0.2%** | **95.36 KB** | **95.21 KB** | **31.18 KB** | **31.14 KB** | **UMD_PROD**
react-dom.development.js | -0.0% | -0.5% | 656.33 KB | 656.12 KB | 153.81 KB | 153.01 KB | NODE_DEV
**react-dom.production.min.js** | **-0.2%** | **-0.3%** | **95.3 KB** | **95.16 KB** | **30.91 KB** | **30.8 KB** | **NODE_PROD**
ReactDOM-dev.js | -0.0% | -0.6% | 663.69 KB | 663.49 KB | 151.92 KB | 151.06 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **-0.1%** | **-0.7%** | **287.74 KB** | **287.38 KB** | **53.32 KB** | **52.92 KB** | **FB_WWW_PROD**
react-dom.profiling.min.js | -0.1% | -0.3% | 98.18 KB | 98.04 KB | 31.51 KB | 31.41 KB | NODE_PROFILING
ReactDOM-profiling.js | -0.1% | -0.8% | 293.15 KB | 292.78 KB | 54.64 KB | 54.21 KB | FB_WWW_PROFILING

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by gaearon** [comment]
>My original PR also removed synchronization for the checked attribute. However it isn't as problematic, and wasn't mentioned in #13525, so I didn't remove the behavior.

I talked to @sebmarkbage, and the guiding principles we came up with are:

* React DOM should do what you'd do if you were writing mutations manually.
* We only support mutations on uncontrolled components.

Our goal is to simplify the system and reduce the edge cases. So if we only set it out of courtesy for e.g. CSS selectors, I think we should stop doing that, and keep it minimal. Just let `value` set `value`, `defaultValue` set `defaultValue`, and same for `checked` / `defaultChecked`. 

Regarding `form.reset()`. We should support it on uncontrolled components. For them, reverting to last supplied `defaultValue` makes sense. Which would happen anyway. We don't need to fire a change event for that. If you use a reset button, put an `onReset` on it if you want.

We should *not* support `form.reset()` on a controlled component. It is a mutation, and if you want it, just use an uncontrolled component — or change your resetting logic to reset the state instead. So we'll want to warn if you reset a form, and it contains a controlled component inside.

**Comment by gaearon** [comment]
You might also want to start by adding a feature flag for this. Because we'll want to keep both versions around for a while.

It might be reasonable to fork some files like `ReactDOMFiberInput` completely later. Let's see which ones have the most feature flag checks after a bit of work.

**Comment by nhunzaker** [comment]
I did some more work today that I believe gets us most of the way there. I shouldn't continue to refine this anymore, next I need to:

- [ ] Break down the changes into distinct feature flags
- [ ] Test a few assumptions about defaultValue since we wrote the Chromium fix for value detachment
- [ ] Follow up on some server side mismatching

I'm really curious about the final item. How should SSR work when `value` has no influence over the value attribute client-side?

I'm also starting to think that the rules are simple enough now that we can start to think about this more as a spec. I think I'd like to write up some rules, even if it's just for myself.

**Comment by nhunzaker** [comment]
One idea for SSR is to very strictly adhere to "value sets the property, defaultValue sets the attribute". For this to work, we'd need to allow a user to set both properties at once without warning.

I don't necessarily sponsor this idea 😸. This would be a breaking change for just about every server-side app, but it would keep things consistent in both environments. 

It also requires a user to be intentional about setting the value attribute for sensitive information, like passwords, which could be a desired side-effect for some users.

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js [hunk 2]
```diff
@@ -1640,8 +1644,16 @@ describe('ReactDOMInput', () => {
...
+      node.focus()
+
       setUntrackedValue.call(node, '2');
       dispatchEventOnNode(node, 'input');
+
+      // TODO: it is unclear why blur must be triggered twice,
+      // manual testing in the fixtures shows that the active element
+      // is no longer the input, however blur() + a blur event seem to
+      // be the only way to remove focus in JSDOM
+      node.blur()
```
Comment (nhunzaker) [inline packages/react-dom/src/__tests__/ReactDOMInput-test.js]: @gaearon this was the false positive case mentioned earlier. JSDOM doesn't update the active element on blur, only after blur() + a blur event. 

I have also walked through the number fixtures to verify that things continue to work as expected manually.

File: packages/react-dom/src/client/ReactDOMInput.js [hunk 3]
```diff
@@ -57,13 +58,14 @@ function isControlled(props) {
 
 export function getHostProps(element: Element, props: Object) {
   const node = ((element: any): InputWithWrapperState);
-  const checked = props.checked;
+  const checked =
+    props.checked != null ? props.checked : node._wrapperState.initialChecked;
```
Comment (gaearon) [inline packages/react-dom/src/client/ReactDOMInput.js]: You're doing `props.checked` access twice now. Let's avoid changes like this.

File: packages/react-dom/src/client/ReactDOMInput.js [hunk 4]
```diff
@@ -212,35 +228,52 @@ export function postMountWrapper(
...
   const node = ((element: any): InputWithWrapperState);
 
+  // Do not assign value if it is already set. This prevents user text input
+  // from being lost during SSR hydration.
   if (props.hasOwnProperty('value') || props.hasOwnProperty('defaultValue')) {
+    const initialValue = node._wrapperState.initialValue;
+    const currentValue = node.value;
+    const value = disableInputAttributeSyncing
+      ? getToStringValue(props.value)
+      : initialValue;
```
Comment (gaearon) [inline packages/react-dom/src/client/ReactDOMInput.js]: Seems like previous we didn't have to call this earlier than next condition. What changed? Let's avoid adding more eager computation.

Comment (nhunzaker) [reply packages/react-dom/src/client/ReactDOMInput.js]: Something can be done here. Working on it.

Comment (nhunzaker) [reply packages/react-dom/src/client/ReactDOMInput.js]: I was! I did this in addffb81f

File: packages/react-dom/src/client/ReactDOMInput.js [hunk 5]
```diff
@@ -212,35 +240,66 @@ export function postMountWrapper(
...
-    if (
-      (type === 'submit' || type === 'reset') &&
-      (props.value === undefined || props.value === null)
-    ) {
+    if (isButton && (props.value === undefined || props.value === null)) {
       return;
     }
 
-    const initialValue = toString(node._wrapperState.initialValue);
-    const currentValue = node.value;
```
Comment (nhunzaker) [inline packages/react-dom/src/client/ReactDOMInput.js]: I inlined this value because it is only needed in one place in both feature flag branches.

**Comment by nhunzaker** [comment]
I believe all comments have been addressed and, reading through the code, I feel good about this change.

File: packages/react-dom/src/client/ReactDOMInput.js [hunk 6]
```diff
@@ -212,35 +240,66 @@ export function postMountWrapper(
...
+          }
+        }
+      } else {
+        // When syncing the value attribute, the value property should use
+        // the the wrapperState._initialValue property. This uses:
+        //
+        //   1. The value React property when present
+        //   2. The defaultValue React property when present
+        //   3. An empty string
+        if (initialValue !== node.value) {
```
Comment (gaearon) [inline packages/react-dom/src/client/ReactDOMInput.js]: This used to compare `toString(node._wrapperState.initialValue) !== node.value`.

But now it compares `node._wrapperState.initialValue !== node.value`.

Is that change safe?

File: packages/react-dom/src/client/ToStringValue.js [hunk 7]
```diff
@@ -18,8 +18,8 @@ export opaque type ToStringValue =
 // Flow does not allow string concatenation of most non-string types. To work
 // around this limitation, we use an opaque type that can only be obtained by
 // passing the value through getToStringValue first.
-export function toString(value: ToStringValue): string {
-  return '' + (value: any);
+export function toString(value: ?ToStringValue): string {
+  return value == null ? '' : '' + (value: any);
```
Comment (gaearon) [inline packages/react-dom/src/client/ToStringValue.js]: Why this change? Did we check every `toString` callsite to ensure it doesn't break them?

File: packages/react-dom/src/client/ReactDOMInput.js (hunk 6 above)
Comment (nhunzaker) [reply packages/react-dom/src/client/ReactDOMInput.js]: You are correct. This should call toString()

File: packages/react-dom/src/client/ToStringValue.js (hunk 7 above)
Comment (nhunzaker) [reply packages/react-dom/src/client/ToStringValue.js]: There were two places in postMount where it could stringify null or undefined. Otherwise this can stay the same. Updating now.

File: packages/react-dom/src/client/ReactDOMInput.js (hunk 6 above)
Comment (gaearon) [reply packages/react-dom/src/client/ReactDOMInput.js]: Needs a regression test?

File: packages/react-dom/src/client/ToStringValue.js (hunk 7 above)
Comment (gaearon) [reply packages/react-dom/src/client/ToStringValue.js]: My priority here is that the old code path stays as similar as it can be. So that we can merge this without worrying.

File: packages/react-dom/src/events/__tests__/ChangeEventPlugin-test.internal.js [hunk 8]
```diff
@@ -87,8 +88,14 @@ describe('ChangeEventPlugin', () => {
     );
     node.dispatchEvent(new Event('input', {bubbles: true, cancelable: true}));
     node.dispatchEvent(new Event('change', {bubbles: true, cancelable: true}));
-    // There should be no React change events because the value stayed the same.
-    expect(called).toBe(0);
+
+    if (ReactFeatureFlags.disableInputAttributeSyncing) {
+      // TODO: figure out why. This might be a bug.
+      expect(called).toBe(1);
```
Comment (nhunzaker) [inline packages/react-dom/src/events/__tests__/ChangeEventPlugin-test.internal.js]: It's possible that this happens because `value` isn't present, so no operation occurs on `value`, so value tracking is never triggered. A solution could be to set the default value or manually call tracking.

File: packages/react-dom/src/client/ToStringValue.js (hunk 7 above)
Comment (nhunzaker) [reply packages/react-dom/src/client/ToStringValue.js]: Totally. I've done this in e4f68fee7

File: packages/react-dom/src/client/ReactDOMInput.js (hunk 6 above)
Comment (nhunzaker) [reply packages/react-dom/src/client/ReactDOMInput.js]: Hard. We only need to call toString for value comparison so that we do not reassign the existing default value and trigger validation. `null` and `undefined` are already skipped. This also happens on the post mount step, and not during hydration. I'm not sure 

We have a manual test fixture for this reason (sorry, I know the formatting isn't ideal):
https://github.com/facebook/react/blob/a7bd7c3c0480c96e32d9518bb707172dfc72db58/fixtures/dom/src/components/fixtures/text-inputs/index.js#L65-L99

**Comment by burka** [comment]
Is there a way to toggle this feature flag without recompiling react? 
When is this feature scheduled?

**Comment by kripod** [comment]
Is there a reason this hasn't been enabled by default? It seems to be a cause for the bug #17609.

**Comment by gaearon** [comment]
No particular reason other than it's a breaking change so we can't enable it until 17.
